src/java.base/share/classes/java/lang/StringUTF16.java

```diff
@@ -158,30 +170,203 @@ public static byte[] toBytes(char[] value, int off, int len) {
         return val;
     }
 
-    public static byte[] compress(char[] val, int off, int len) {
-        byte[] ret = new byte[len];
-        if (compress(val, off, ret, 0, len) == len) {
-            return ret;
+    /**
+     * {@return Compress the char array (containing UTF16) into a compact strings byte array}
+     * If all the chars are LATIN1, it returns an array with len == count,
+     * otherwise, it contains UTF16 characters.
+     * <p>
+     * A UTF16 array is returned *only* if at least 1 non-latin1 character is present.
+     * This must be true even if the input array is modified while this method is executing.
+     * This is assured by copying the characters while checking for latin1.
+     * If all characters are latin1, a byte array with length equals count is returned,
+     * indicating all latin1 chars. The scan may be implemented as an intrinsic,
+     * which returns the index of the first non-latin1 character.
+     * When the first non-latin1 character is found, it switches to creating a new
+     * buffer; the saved prefix of latin1 characters is copied to the new buffer;
+     * and the remaining input characters are copied to the buffer.
+     * The index of the known non-latin1 character is checked, if it is latin1,
+     * the input has been changed. In this case, a second attempt is made to compress to
+     * latin1 from the copy made in the first pass to the originally allocated latin1 buffer.
+     * If it succeeds the return value is latin1, otherwise, the utf16 value is returned.
+     * In this unusual case, the result is correct for the snapshot of the value.
+     * The resulting string may be indeterminate if the input array is modified during this
+     * operation, but it is ensured that at least 1 non-latin1 character is present in
+     * the non-latin1 buffer.
+     *
+     * @param val   a char array
+     * @param off   starting offset
+     * @param count length of chars to be compressed, length > 0
+     */
+    @ForceInline
+    public static byte[] compress(final char[] val, final int off, final int count) {
+        byte[] latin1 = new byte[count];
```

Comment:
Will this redundant array allocation be costly if we are working with mostly-utf16 strings, such as CJK strings with no latin characters?

I suggest we can use a heuristic to read the initial char; if it's utf16 then we skip the latin-1 process altogether (and we can assign the utf16 value to the initial index to ensure it's non-latin-1 compressible.

Reply:
Just curious, how does benchmark StringConstructor.newStringFromCharsMixedBegin change before and after this patch? If we can see how much of an impact this has on CJK strings it would be appreciated.